bokehjs/test/integration/plots.ts

```diff
@@ -107,6 +107,112 @@ describe("Plot", () => {
     })
   })
 
+  describe("should support windowed auto-ranging", async () => {
+    function plot(window_axis: WindowAxis) {
+      const r = new Range1d({start: 0, end: 6})
+      const dr = new DataRange1d({range_padding: 0})
+      const p = figure({
+        sizing_mode: "fixed",
+        width: 300,
+        height: 300,
+        window_axis,
+        title: `window_axis == ${window_axis}`,
+        x_range: window_axis == "x" ? r: dr,
+        y_range: window_axis == "x" ? dr: r,
+      })
+      const s = p.scatter({x: [1, 2, 3, 4, 5], y: [1, 2, 3, 4, 5], size: 8})
+      return {p, s, r, dr}
+    }
+    for (const wax of ["x", "y"]) {
+      it(`with window_axis='${wax}' when data changes`, async () => {
+        const nwax = (wax == "x") ? "y" : "x"
+        const {p, s, dr} = plot(wax as WindowAxis)
+        const {view} = await display(p)
+
+        expect(dr.start).to.be.equal(1)
+        expect(dr.end).to.be.equal(5)
+
+        // updated data in fixed range
+        const data: any = {}
+        data[wax] = [1, 2, 3, 4, 5]
+        data[nwax] = [1, 2, 10, 4, 5]
+        s.data_source.data = data
+        await view.ready
+        await view.ready
+        expect(dr.start).to.be.equal(1)
+        expect(dr.end).to.be.equal(10)
+
+        // updated data not in fixed range (no change)
+        const data2: any = {}
+        data2[wax] = [-1, 1, 2, 3, 4, 100]
+        data2[nwax] = [-100, 1, 2, 10, 4, 100]
+        s.data_source.data = data2
+        await view.ready
+        await view.ready
+        expect(dr.start).to.be.equal(1)
+        expect(dr.end).to.be.equal(10)
+
+        // streamed data in fixed range
+        const new_data: any = {}
+        new_data[wax] = [2.5, 3.5]
+        new_data[nwax] = [-90, 100]
+        s.data_source.stream(new_data)
+        await view.ready
+        await view.ready
+        expect(dr.start).to.be.equal(-90)
+        expect(dr.end).to.be.equal(100)
+
+        // streamed data not in fixed range (no change)
+        const new_data2: any = {}
+        new_data2[wax] = [-1]
+        new_data2[nwax] = [200]
+        s.data_source.stream(new_data2)
+        await view.ready
+        await view.ready
+        expect(dr.start).to.be.equal(-90)
+        expect(dr.end).to.be.equal(100)
+
+        // new negative data in fixed range
+        const data3: any = {}
+        data3[wax] = [1, 2, 3, 4, 5]
+        data3[nwax] = [1, 2, 10, 4, -10]
+        s.data_source.data = data3
+        await view.ready
+        await view.ready
+        expect(dr.start).to.be.equal(-10)
+        expect(dr.end).to.be.equal(10)
+
+        await view.ready
+      })
+    }
+    for (const wax of ["x", "y"]) {
+      it(`with window_axis='${wax}' when range changes`, async () => {
+        const {p, r, dr} = plot(wax as WindowAxis)
```

Comment:
You can get rid of the cast with:
```suggestion
    for (const wax of ["x", "y"] as const) {
      it(`with window_axis='${wax}' when range changes`, async () => {
        const {p, r, dr} = plot(wax)
```
Same in other cases.

Reply:
https://github.com/bokeh/bokeh/pull/14353/commits/bb9f114897c1b9fc9dae637fb8460d386c189179